tfjs-layers/src/layers/nlp/tokenizers.ts

```diff
@@ -134,24 +136,338 @@ export abstract class Tokenizer extends Layer {
+    const minPairRankIndices = maskedPairRanks.map(arr => argMin(arr));
+
+    // Get words and pairs to process.
+    const unfinishedWords = wordsStr.filter((_, idx) => mask[idx]);
+
+    const pairLeft = unfinishedWords.map(
+      (word, idx) => word[minPairRankIndices[idx]]);
+
+    const pairRight = unfinishedWords.map(
+      (word, idx) => word[minPairRankIndices[idx] + 1]);
+
+    const mergedPairs = pairLeft.map((left, idx) => {
+      const right = pairRight[idx];
+      return [left, right].join('');
+    });
+    const unfinishedWordsIndices = mask
+      .map((_, idx) => idx)
+      .filter((_, idx) => mask[idx]);
+
+    const mergedPairIndices = unfinishedWordsIndices.map(
+      (index, idx) => [index, minPairRankIndices[idx]]);
+    const emptyStringIndices = unfinishedWordsIndices.map(
+      (index, idx) => [index, minPairRankIndices[idx] + 1]);
+
+    mergedPairIndices.forEach((indices, idx) => {
+      const [wordIdx, charIdx] = indices;
+      const mergedPair = mergedPairs[idx];
+      wordsStr[wordIdx][charIdx] = mergedPair;
+    });
+
+    emptyStringIndices.forEach((indices) => {
+      const [wordIdx, charIdx] = indices;
+      wordsStr[wordIdx][charIdx] = '';
+    });
+
+    words = wordsStr.map(word => tensor(word));
+    words = await removeStringsFromInputs(words, '');
+
+    return [words, mask];
+  }
+
+  private async bpeMerge(words: Tensor[]): Promise<Tensor[]> {
+    const numWords = words.length;
+
+    // Merge bytes.
+    function loopCondition(mask: boolean[]): boolean {
+      return mask.some(e => e);
+    }
+
+    const initialMask: boolean[] = Array(numWords).fill(true);
+
+    let mergedWords = words;
+    let mask = initialMask;
+    while (loopCondition(mask)) {
+      [mergedWords, mask] = await this.bpeMergeOneStep(mergedWords, mask);
+    }
+
+    return mergedWords;
+  }
+
+  /**
+   * Map token bytes to unicode using `byte2unicode`.
+   */
+  private async transformBytes(tokens: Tensor): Promise<Tensor[]> {
+    const tokensStr = await tokens.data() as unknown as string[];
+
+    const splitBytes = tokensStr.map(
+      token => tensor(token.split('').map(c => c.charCodeAt(0))));
+    const splitUnicode = await this.byte2Unicode.lookup(splitBytes);
+
+    return splitUnicode;
+  }
+
+  /**
+   * Process unseen tokens and add to cache.
+   */
+  private async bpeMergeAndUpdateCache(tokens: Tensor) {
+    const words = await this.transformBytes(tokens);
+    const tokenizedWordsTensor = await this.bpeMerge(words);
+    const tokenizedWords =
+      await tensorArrTo2DArr<string>(tokenizedWordsTensor);
+
+    // For each word, join all its token by a whitespace,
+    // e.g., ["dragon", "fly"] => "dragon fly" for hash purpose.
+    const joinedTokens = tokenizedWords.map(word => word.join(' '));
+
+    await this.cache.insert(tokens, joinedTokens);
+  }
+
+  tokenize(inputs: Tensor): Tensor[] {
+    this.bpeMergeAndUpdateCache;
```

Comment:
Because these private methods aren't used until the tokenizer() method is implemented, I was getting lint errors. For now, I've simply included them in the NotImplemented error message.